core/debugger/remote_debugger_peer.cpp

```diff
@@ -89,64 +97,121 @@ RemoteDebuggerPeerTCP::RemoteDebuggerPeerTCP(Ref<StreamPeerTCP> p_tcp) {
+					continue;
+				}
+				current_write_channel = channel_index;
+				break;
 			}
-			mutex.lock();
-			Variant var = out_queue[0];
-			out_queue.pop_front();
-			mutex.unlock();
+		}
+		if (current_write_channel < 0) {
+			// Nothing to do.
+			return;
+		}
+
+		uint8_t *buf = channels[current_write_channel].out_buf.ptrw();
+
+		if (channels[current_write_channel].out_left <= 0) {
+			// Need to start next buffer
+			channels[current_write_channel].shared.mutex.lock();
+			Variant var = channels[current_write_channel].shared.out_queue[0];
+			channels[current_write_channel].shared.out_queue.pop_front();
+			channels[current_write_channel].shared.mutex.unlock();
 			int size = 0;
+			const int OVERHEAD = 4;
 			Error err = encode_variant(var, nullptr, size);
-			ERR_CONTINUE(err != OK || size > out_buf.size() - 4); // 4 bytes separator.
-			encode_uint32(size, buf);
-			encode_variant(var, buf + 4, size);
-			out_left = size + 4;
-			out_pos = 0;
+			if (err != OK || size > channels[current_write_channel].out_buf.size() - OVERHEAD) {
+				// Can't send, but we did service this item.
+				WARN_PRINT(vformat("Failed to send debugger message to channel %d; error %d.", current_write_channel, err));
+				current_write_channel = -1;
+				continue;
+			}
+			static_assert(NUM_CHANNELS < 256);
+			static_assert(MAX_MESSAGE_SIZE < (1 << 24));
+			encode_uint32(size | (current_write_channel << 24), buf);
+			encode_variant(var, buf + OVERHEAD, size);
+			channels[current_write_channel].out_left = size + OVERHEAD;
+			channels[current_write_channel].out_pos = 0;
 		}
+
 		int sent = 0;
-		tcp_client->put_partial_data(buf + out_pos, out_left, sent);
-		out_left -= sent;
-		out_pos += sent;
+		tcp_client->put_partial_data(buf + channels[current_write_channel].out_pos, channels[current_write_channel].out_left, sent);
+		channels[current_write_channel].out_left -= sent;
+		channels[current_write_channel].out_pos += sent;
+
+		if (channels[current_write_channel].out_left <= 0) {
+			// Done with the current transmission.
+			current_write_channel = -1;
+		}
 	}
 }
 
 void RemoteDebuggerPeerTCP::_read_in() {
 	while (tcp_client->get_status() == StreamPeerTCP::STATUS_CONNECTED && tcp_client->wait(NetSocket::POLL_TYPE_IN) == OK) {
-		uint8_t *buf = in_buf.ptrw();
-		if (in_left <= 0) {
-			if (in_queue.size() > max_queued_messages) {
-				break; // Too many messages already in queue.
-			}
-			if (tcp_client->get_available_bytes() < 4) {
-				break; // Need 4 more bytes.
+		if (current_read_channel < 0) {
+			// Could be reading any channel, so don't do it if any of them are blocked.
+			// REVISIT: this isn't completely correct.  If Main is blocked on thread.join, it will never respond to any
+			// requests and we need to give up, so we do need per-channel flow control, i.e. multiple sockets or our
+			// own flow control (e.g. enet) or just conceptually tear down the Main connection and discard all Main
+			// messages.
```

Comment:
is there a reason why we can't:
- Always poll events from main.
- Only poll events from threads while they are debug-broken (or always poll but never pop messages unless debugging).
- When debug broken:
- If inside a thread, drop all messages that are not "core:" (which is the debugging part, or a list of "thread-safe" prefix)
- If inside main, process all messages regularly.
else:
- Process all messages from main (if main is blocked, and no thread is debugging, messages will simply queue up until main returns available or a thread reaches a breakpoint/error).

We could also then (in a separate PR) further tap into Thread/Mutex/Semaphore to poll_events from main at a regular basis in debug builds (like we do for infinite loops in scripts).

Reply:
If you service a queue from more than one thread, messages can be processed out of order.  I felt that would be a weaker mechanism than what I built, so I would only choose that if it buys me something significant in return.   I don't fully understand what the benefit would be of doing it the way that you suggest.  I think it would generally work though, yes.

Right now, the broken thread (focused thread) does the debugging related core messages in order, and main does the other stuff when it can.    It isn't perfect for sure.  For example, when you try to use the remote scene debugger, usually main isn't available until you step execute, which pretty much sucks.  If we wanted to make that nicer though, we would have to do some sort of explicit handshaking with main where the debug thread wakes up main just to service captures but main isn't otherwise allowed to run the main loop so the game doesn't progress.   That would be an evolution I could see making sense, but it is a bit harder to get right.

Reply:
Actually, scratch that last part... you just explained that above and I just didn't get it on first read.  Main could totally spin on the thread barrier and still process debug messages while another thread is broken.   Threads other than main would go block when "held" as they do now but main would be "held" by spinning in a loop that processes the main queue.  Then the only time the UI stops updating the remote scene debugger would be if main is REALLY blocked (OS semaphore, mutex, etc.)

That would be a great enhancement, I think.

Reply:
I went back and tested the behavior of *Main* when it is held in GDScript code, and yes it was a bug.   It just blocked on the "held threads" barrier and didn't process things, so remote inspection generally didn't work when debugging non-main threads (unless *Main* was idle.)

I implemented the change outlined above to let *Main* spin and process messages but nothing else, as soon as it executes in the script engine(s).  In a "normal" program, it will immediately enter this state when another thread is being step executed.   That fixes being able to remote inspect nodes while debugging other threads.

https://github.com/godotengine/godot/compare/2b55d9f4b2ddc2b13d5977a09177f96390ab5908..217f9316f353f23c8517ef43a22374ce53469002